### PR TITLE
VIS: Fix DataFrame.plot() produces incorrect legend markers when plotting multiple series on the same axis

### LINKED ISSUE #18222: DataFrame.plot() produces incorrect legend labels when plotting multiple series on the same axis
#### Code Sample, a copy-pastable example if possible
```python
import pandas as pd
import matplotlib.pyplot as plt
df = pd.DataFrame(data=[[1, 1, 1, 1], [2, 2, 4, 8]], columns=['x', 'r', 'g', 'b'])
fig, ax = plt.subplots(nrows=1, ncols=3)
# Left plot
df.plot(x='x', y='r', linewidth=0, marker='o', color='r', ax=ax[0])
df.plot(x='x', y='g', linewidth=1, marker='x', color='g', ax=ax[0])
df.plot(x='x', y='b', linewidth=1, marker='o', color='b', ax=ax[0])
# Center plot
df.plot(x='x', y='b', linewidth=1, marker='o', color='b', ax=ax[1])
df.plot(x='x', y='r', linewidth=0, marker='o', color='r', ax=ax[1])
df.plot(x='x', y='g', linewidth=1, marker='x', color='g', ax=ax[1])
# Right plot
df.plot(x='x', y='g', linewidth=1, marker='x', color='g', ax=ax[2])
df.plot(x='x', y='b', linewidth=1, marker='o', color='b', ax=ax[2])
df.plot(x='x', y='r', linewidth=0, marker='o', color='r', ax=ax[2])
# Produces correct output when uncommented:
# ax[0].legend()
# ax[1].legend()
# ax[2].legend()
plt.show()
```

#### Problem description
**Note:** Possibly related to #14958, #17939, #14563, however this issue discusses how the behaviour depends on the order in which `df.plot()` is called

- In certain situations, `df.plot()` may generate incorrect legend labels (see example)
- Incorrect legend labels may appear when `df.plot()` plots multiple series on the same axis
- Plotting only one series per axis always appears to produce the correct legend label
- When multiple series are plotted on the same axis:
    - Only the last `df.plot()` call will produce the correct legend label
    - Calling `ax.legend()` after the final `df.plot()` will produce the correct legend entry
- Calling `df.plot()` with `markerstyle=''` always appears produces the correct legend label for that series regardless of the order in which `df.plot()` is called, or the number of other series on the same axis

#### Expected Output
**This is the incorrect output produced by the sample code**
Note that in each plot, only the last series to be plotted has a correct legend label
image
**This is the expected output**
image

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.4.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-37-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: None
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: None
numpy: 1.13.3
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 6.1.0
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
I found out that the reason this happens is because `subplots` is not updated accordingly when `plt.subplots()` is used and assign to `ax`, e.g. `fig, ax = plt.subplots(nrows=1, ncols=3)`. I tested my solution and things all get solved, and figure looks very correct like below:
Screen Shot 2019-08-07 at 10 03 39 PM

BTW, i am not very sure how to properly test this, and would be nice to know!

- [x] closes #18222
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by charlesdong1991** [comment]
Screen Shot 2019-08-07 at 10 03 39 PM
here is how plotting looks like now, and i am not sure how to properly test it, would be nice to know!

**Review by TomAugspurger — COMMENTED** [review]
A test would be nice. Does the issue post have one?

File: pandas/plotting/_matplotlib/core.py [hunk 1]
```diff
@@ -166,6 +166,15 @@ def __init__(
             setattr(self, attr, value)
 
         self.ax = ax
+
+        # set subplots to True if plt.subplots() uses and assign to ax
+        from matplotlib.axes import SubplotBase
+        try:
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: yeah, i moved this part to `_make_legend`, and also removed this confusing part. Seems could still correctly plot this case

File: pandas/plotting/_matplotlib/core.py [hunk 2]
```diff
@@ -166,6 +166,15 @@ def __init__(
             setattr(self, attr, value)
 
         self.ax = ax
+
+        # set subplots to True if plt.subplots() uses and assign to ax
+        from matplotlib.axes import SubplotBase
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/core.py]: moved!

**Comment by charlesdong1991** [comment]
@TomAugspurger thanks very much for your quick review! Just curious about what you mean by `Does the issue post have one?` this issue does provide a fail case, but no test is given. 
I do see some tests for plotting, but since this failure is mainly caused by missing markers in legend, haven't found a proper solution to test if this marker/color is printed correctly. Would be very nice if you could give some hint! ^^

**Comment by TomAugspurger** [comment]
Ideally we would make a plot and verify that the legend labels and markers
are matched up. I can never remember the matplotlib API for that though.

**Comment by charlesdong1991** [comment]
Thanks for your quick reply, I will check it up @TomAugspurger

**Comment by charlesdong1991** [comment]
This is kind of new working solution with a test case that is passing all tests on my local computer, not sure if it can also pass all tests here. But in the meantime pls feel free to take a look and provide reviews if you have some time. @TomAugspurger

**Comment by charlesdong1991** [comment]
any follow-up reviews? all are much appreciated! ^^

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Looks good. I'll restart the windows azure CI failure.

**Comment by charlesdong1991** [comment]
ping

**Comment by TomAugspurger** [comment]
Thanks!

**Comment by jzlcdh** [comment]
If you need another example of this problem let me know

**Pull request merged**